modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/ArgumentParser.java

```diff
@@ -0,0 +1,514 @@
+        arg(
+            CMD_ENABLE_EXPERIMENTAL,
+            CMD_ENABLE_EXPERIMENTAL, Boolean.class,
+            IgniteSystemProperties.getBoolean(IGNITE_ENABLE_EXPERIMENTAL_COMMAND)
+        );
+    }
+
+    /** */
+    private <T> void arg(String name, String usage, Class<T> type, T dflt, BiConsumer<String, T> validator) {
+        common.add(optionalArg(name, usage, type, t -> dflt, validator));
+    }
+
+    /** */
+    private <T> void arg(String name, String usage, Class<T> type, T dflt) {
+        common.add(optionalArg(name, usage, type, () -> dflt));
+    }
+
+    /**
+     * Creates list of common utility options.
+     *
+     * @return Array of common utility options.
+     */
+    public String[] getCommonOptions() {
+        List<String> list = new ArrayList<>();
+
+        for (CLIArgument<?> arg : common) {
+            if (arg.name().equals(CMD_AUTO_CONFIRMATION))
+                continue;
+
+            if (isBoolean(arg.type()))
+                list.add(asOptional(arg.name(), true));
+            else
+                list.add(asOptional(arg.name() + " " + arg.usage(), true));
+        }
+
+        return list.toArray(U.EMPTY_STRS);
+    }
+
+    /**
+     * Parses and validates arguments.
+     *
+     * @param raw Raw arguments.
+     * @return Arguments bean.
+     * @throws IllegalArgumentException In case arguments aren't valid.
+     */
+    <A extends IgniteDataTransferObject> ConnectionAndSslParameters<A> parseAndValidate(
+        List<String> raw
+    ) {
+        List<String> args = new ArrayList<>(raw);
+
+        IgniteBiTuple<Command<A, ?>, Command<?, ?>> cmd = command(args.iterator());
+
+        IgniteBiTuple<List<CLIArgument<?>>, List<CLIArgument<?>>> cmdArgs = commandArguments(cmd.get1());
+
+        cmdArgs.get2().addAll(common);
+
+        CLIArgumentParser parser = new CLIArgumentParser(cmdArgs.get1(), cmdArgs.get2());
+
+        parser.parse(args.iterator());
+
+        A arg = argument(
+            cmd.get1().argClass(),
+            (fld, pos) -> parser.get(pos),
+            fld -> parser.get(toFormattedFieldName(fld).toLowerCase())
+        );
+
+        if (!parser.<Boolean>get(CMD_ENABLE_EXPERIMENTAL) && cmd.get1().getClass().isAnnotationPresent(IgniteExperimental.class)) {
+            log.warning(
+                String.format("To use experimental command add " + CMD_ENABLE_EXPERIMENTAL + " parameter for %s",
+                    UTILITY_NAME)
+            );
+
+            throw new IllegalArgumentException("Experimental commands disabled");
+        }
+
+        return new ConnectionAndSslParameters<>(cmd.get1(), cmd.get2(), arg, parser);
+    }
+
+    /**
+     * Get command from hierarchical root.
+     *
+     * @param iter Iterator of commands names.
+     * @return Command to execute.
+     * @param <A> Argument type.
+     */
+    protected <A extends IgniteDataTransferObject> IgniteBiTuple<Command<A, ?>, Command<?, ?>> command(
```

Comment:
Here and below. It seems that using just tuple makes code a little bit hard to understand. What do you think about creating a small separate class for this task?